src/performance/scala-2.11/markets/MarketRegulatorActor.scala

```diff
@@ -17,42 +17,47 @@ package markets
 
 import akka.actor.{ActorRef, PoisonPill, Props, Terminated}
 
-import markets.tradables.Tradable
 
-import scala.collection.mutable
+/** Class representing a market regulatory agency.
+  *
+  * @param participants
+  * @param markets
+  * @note The `MarketRegulatorActor` supervises a collection of market participants as well as a
+  *       collection of markets.  When there are no longer any market participants, the
+  *       `MarketRegulatorActor` shutdowns all the markets and terminates the actor system.
+  */
+class MarketRegulatorActor(participants: Iterable[ActorRef],
+                           markets: Iterable[ActorRef]) extends StackableActor {
 
-
-class MarketRegulator(brokerage: ActorRef,
-                      markets: mutable.Map[Tradable, ActorRef]) extends StackableActor {
+  participants.foreach(participant => context.watch(participant))
+  markets.foreach(market => context.watch(market))
 
   wrappedBecome(marketRegulatorBehavior)
 
-  /** A MarketRegulator maintains a list of market participants that it monitors. */
-  val regulatedEntities = mutable.Set.empty[ActorRef]
-
-  // MarketRegulator monitors both the brokerage and the markets...
-  context.watch(brokerage); regulatedEntities += brokerage
-  markets.foreach{ case (_, market) => context.watch(market); regulatedEntities += market }
-
   def marketRegulatorBehavior: Receive = {
-    case Terminated(entity) =>
-      regulatedEntities -= entity
-      if (entity == brokerage) {
-        markets.foreach { case (_, market) => market tell(PoisonPill, self) }
-      } else if (regulatedEntities.isEmpty) {
+    case Terminated(entity) if _participants.contains(entity) =>
+      _participants -= entity
+      if (_participants.isEmpty) {
+        _markets.foreach(market => market tell(PoisonPill, self))
+      }
+    case Terminated(entity) if _markets.contains(entity) =>
+      _markets -= entity
+      if (_markets.isEmpty) {
         context.system.terminate()
-      } else {
-        // do nothing!
       }
   }
 
+  // Internally represent regulated entities as Sets
+  private var _participants = participants.toSet
+  private var _markets = markets.toSet
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [List(_markets) should have a smaller scope by using private[this]](https://www.codacy.com/app/davidrpugh/markets-sandbox/file/2444281914/issues/source?bid=2921361&fileBranchId=3246024#l52)